### PR TITLE
Porting cubemap compute shaders to raster for the mobile renderer

### PR DESCRIPTION
**Pull request by BastiaanOlij**
This PR implement raster versions of the effect shaders related to our cubemaps (for use with sky and reflection probes) as on mobile hardware the computer versions aren't reliable and won't work on most devices.

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
This now works for background sky if the following settings is off:
image

Result so far:
image
*note* The lines/artifacts are because of issues with shadow mapping on spotlights unrelated to this PR.

Still todo:
- implement downgrade shader so we can make this work if the above setting is on
- check other paths with different backgrounds as different shaders are used if no sky is used or if sky is updated every frame
- fix this for reflection probes
- fix compiler issues on different platforms 

Without the downgrade shader... yuck :P
image

**Comment by BastiaanOlij** [comment]
Downgrade shader now works in fragment mode, and `cubemap_filter` which is used as a faster less precise method for real time updating the sky (and thus its radiance maps). The problem here is that if you're using array mode, all the roughness levels get created the same. There is a little extra happening in the compute shader that I've not applied to the fragment version. Possibly I can cheat and just increase the mipmap base level for each roughness level (or artificially decreasing the size I pass). Will experiment more with this tomorrow.

File: servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp [hunk 1]
```diff
@@ -398,55 +402,139 @@ void RendererSceneSkyRD::ReflectionData::update_reflection_data(int p_size, int
...
-		for (int i = 1; i < layers.size(); i++) {
-			views.push_back(layers[i].views[0]);
+		for (int i = 1; i < downsampled_layer.mipmaps.size(); i++) {
+			for (int k = 0; k < 6; k++) {
+				effects->cubemap_downsample_raster(downsampled_layer.mipmaps[i - 1].view, downsampled_layer.mipmaps[i].framebuffers[k], k, downsampled_layer.mipmaps[i].size);
+			}
+		}
+
+		if (p_use_arrays) {
+			for (int i = 0; i < layers.size(); i++) {
```
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp:439]: The code here does not work properly, this needs to be solved differently or I'm missing something in the fragment shader that allows it to behave the same as its compute counterpart. 

Anyway, with realtime updating the sky, roughness set on our material and using reflection arrays we get this ugly result:
image

It's supposed to look something like this:
image

Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp:439]: It actually doesn't look great with reflection arrays turned off either:
image

So its definitely something I'm not implementing right in the downsample shader.

Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp:439]: This is the sky shader I'm using to test this:
```
shader_type sky;

uniform sampler2D source_panorama : filter_linear;

void sky() {
	vec2 uv = SKY_COORDS;
	uv.x += TIME * 0.01;
	COLOR = texture(source_panorama, uv).rgb;
}
```

Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp:439]: Final reference, this is what it looks like using the clustered shader:
image

**Comment by BastiaanOlij** [comment]
I think I've covered everything with the footnote that I'm not happy about the real time radiance map update logic. Could use some more eyes on this :)

File: servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp [hunk 2]
```diff
@@ -398,55 +402,139 @@ void RendererSceneSkyRD::ReflectionData::update_reflection_data(int p_size, int
...
+				for (int j = 0; j < layers[i].mipmaps.size(); j++) {
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/renderer_scene_sky_rd.cpp:440]: This should not loop through the mip levels. It should only render to mip level 0. 

You can remove this for loop and replace all references to ``j`` with ``0``

I will make a PR to your branch removing this and adding some labels to help with debugging

File: servers/rendering/renderer_rd/shaders/cubemap_filter_raster.glsl [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+					float sample_level = coeffsLevel[iSubTap];
+
+					float sample_weight = coeffsWeight[iSubTap];
+#endif
+
+					sample_weight *= frameweight;
+
+					// adjust for jacobian
+					sample_dir /= max(abs(sample_dir[0]), max(abs(sample_dir[1]), abs(sample_dir[2])));
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/cubemap_filter_raster.glsl:240]: I think this is where the issue is coming from. The coefficients are likely baked assuming the range of values in the compute version. So when you are running this in raster, the samples are getting thrown off.

**Review by clayjohn — COMMENTED** [review]
I have done some work to track down where the bug is coming from. I have a PR that gets you part way to a solution, incoming shortly

**Review by reduz — APPROVED** [review]

**Comment by BastiaanOlij** [comment]
@akien-mga @clayjohn , on godot chat discussed this PR with Reduz. If you are both happy with it we should merge this.

There is still the issue around real time cube map filtering but I think we should raise this as an issue and deal with in a separate PR. It'll likely be a fix we need to do in the shader itself while the rest of the work is sound and seeing this is a feature unlikely to be used on mobile anyway it should probably not hold up merging. We should still try and resolve it :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**